apps/keyboard/js/imes/emoji/emoji.js

```diff
@@ -0,0 +1,90 @@
+/* global InputMethods, PromiseStorage*/
+
+
+'use strict';
+
+(function() {
+  function RecentKeys() {
+    this._items = [];
+  }
+
+  RecentKeys.prototype.RECENT_KEY_COUNT = 18;
+
+  RecentKeys.prototype.getItems = function() {
+    return this._items || [];
+  };
+
+  RecentKeys.prototype.setItems = function(val) {
+    if (Array.isArray(val) && val.length) {
+      this._items = val.slice(0, this.RECENT_KEY_COUNT);
+    }
+  };
+
+  RecentKeys.prototype.getPanelKeys = function() {
+    return this._items.map(function(val) {
+      return { compositeKey: val, type: 'emoji' };
+    });
+  };
+
+  RecentKeys.prototype.getPanelKeysPromise = function() {
```

Comment:
You don't need this anymore because `_items` is always available once `EmojiInputMethod#init` has resolved.